### PR TITLE
feat: lazy validator

### PR DESCRIPTION
**Pull request by imranbarbhuiya**
ref: https://github.com/jquense/yup#lazyvalue-any--schema-lazy

### CODE REVIEW

**Comment by vladfrangu** [comment]
Can you also try
```ts
const obj = s.object({
	id: s.string,
	items: s.lazy(() => obj)
});
```
? Does this work?

**Comment by imranbarbhuiya** [comment]
missing github reporter in vitest xD

**Review by favna — DISMISSED** [review]

File: tests/validators/lazy.test.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	test('GIVEN an invalid value THEN throw ValidationError', () => {
+		expectError(
+			() => predicate.parse('Sapphire'),
+			new ExpectedConstraintError('s.string.lengthEqual', 'Invalid string length', 'Sapphire', 'expected.length === 5')
+		);
+	});
+});
+
+describe('CircularLazyValidator', () => {
+	// @ts-expect-error (circular)
```
Comment (vladfrangu) [inline tests/validators/lazy.test.ts:41]: This shouldn't error tho... Concerning that it is erroring

Comment (favna) [reply tests/validators/lazy.test.ts]: Fixed

**Review by favna — APPROVED** [review]

File: src/validators/LazyValidator.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+		super(constraints);
+		this.validator = validator;
+	}
+
+	protected override clone(): this {
+		return Reflect.construct(this.constructor, [this.validator, this.constraints]);
+	}
+
+	protected handle(values: unknown): Result<R, ValidatorError> {
+		if (!this.shouldRunConstraints) {
```
Comment (vladfrangu) [inline src/validators/LazyValidator.ts:18]: We should always call this.validator.run

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Review by favna — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

File: tests/validators/lazy.test.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+	test('GIVEN circular schema THEN throw ', () => {
+		expectError(
+			() => predicate.parse({ id: 'Hello', items: { id: 'Hello', items: { id: 'Hello' } } }),
+			new CombinedPropertyError([
+				['items', new CombinedPropertyError([['items', new CombinedPropertyError([['items', new MissingPropertyError('items')]])]])]
+			])
+		);
+	});
+});
```
Comment (kyranet) [inline tests/validators/lazy.test.ts:59]: Please make another test but with `items` optional, so we can have a passing test and verify whether the result is correct or not.

**Review by kyranet — CHANGES_REQUESTED** [review]

File: tests/validators/lazy.test.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+	const predicate: SchemaOf<PredicateSchema> = s.object({
+		id: s.string,
+		items: s.lazy<SchemaOf<PredicateSchema>>(() => predicate).optional
+	});
+
+	test('GIVEN circular schema THEN throw ', () => {
+		expect(predicate.parse({ id: 'Hello', items: { id: 'Hello', items: { id: 'Hello' } } })).toStrictEqual({
+			id: 'Hello',
+			items: { id: 'Hello', items: { id: 'Hello' } }
```
Comment (kyranet) [inline tests/validators/lazy.test.ts:75]: oh-dear-gordon-ramsay

**Review by kyranet — APPROVED** [review]

File: tests/validators/lazy.test.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+		id: string;
+		items?: PredicateSchema;
+	}
+
+	const predicate: SchemaOf<PredicateSchema> = s.object({
+		id: s.string,
+		items: s.lazy<SchemaOf<PredicateSchema>>(() => predicate).optional
+	});
+
+	test('GIVEN circular schema THEN throw ', () => {
```
Comment (kyranet) [inline tests/validators/lazy.test.ts:72]: throw?

File: tests/validators/lazy.test.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		expect(predicate.parse({ id: 'Hello', items: { id: 'Hello', items: { id: 'Hello' } } })).toStrictEqual({
```
Comment (kyranet) [inline tests/validators/lazy.test.ts:73]: Please never test with identical values, since it can miss extreme bugs where state is shared.

**Review by favna — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**